influxdb3_client/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+use std::string::FromUtf8Error;
+
+use bytes::Bytes;
+use reqwest::{Body, IntoUrl, StatusCode};
+use secrecy::{ExposeSecret, Secret};
+use serde::Serialize;
+use url::Url;
+
+/// Primary error type for the [`Client`]
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("base URL error: {0}")]
+    BaseUrl(#[source] reqwest::Error),
+
+    #[error("request URL error: {0}")]
+    RequestUrl(#[from] url::ParseError),
+
+    #[error("failed to send /api/v3/write_lp request: {0}")]
+    WriteLpSend(#[source] reqwest::Error),
+
+    #[error("failed to read the API response bytes: {0}")]
+    Bytes(#[source] reqwest::Error),
+
+    #[error("failed to send /api/v3/query_sql request: {0}")]
+    QuerySqlSend(#[source] reqwest::Error),
+
+    #[error("invalid UTF8 in response: {0}")]
+    InvalidUtf8(#[from] FromUtf8Error),
+
+    #[error("server responded with error [{code}]: {message}")]
+    ApiError { code: StatusCode, message: String },
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// The InfluxDB 3.0 Client
+///
+/// For programmatic access to the HTTP API of InfluxDB 3.0
+#[derive(Debug, Clone)]
+pub struct Client {
+    /// The base URL for making requests to a running InfluxDB 3.0 server
+    base_url: Url,
+    /// The `Bearer` token to use for authenticating on each request to the server
+    auth_header: Option<Secret<String>>,
+    /// A [`reqwest::Client`] for handling HTTP requests
+    http_client: reqwest::Client,
+}
+
+impl Client {
+    /// Create a new [`Client`]
+    ///
+    /// A [`reqwest::Client`] is passed in order to allow for re-use of an existing
+    /// `reqwest::Client`.
```

Comment:
Could also have a `Default` implementation or similar that instantiates the `reqwest::Client` internally without having to pass it.

Reply:
I think the difficulty would be that you do need a base url and defaulting to localhost is only really desirable for well local testing. One thing you could do would be a new and new_with_client where new makes a default version and new_with_client would let the user customize their request client for their needs.

Reply:
I will change it to this approach. For now since we are only using this as a one-off in the CLI, we do not need the ability to re-use an existing reqwest client. As per [`reqwest` docs](https://docs.rs/reqwest/latest/reqwest/struct.Client.html), the reason for doing so is:
> The Client holds a connection pool internally, so it is advised that you create one and reuse it.